table/block_based/block.cc

```diff
@@ -243,16 +243,55 @@ void DataBlockIter::Seek(const Slice& target) {
     return;
   }
   uint32_t index = 0;
-  bool ok = BinarySeek<DecodeKey>(seek_key, 0, num_restarts_ - 1, &index,
-                                  comparator_);
+  bool ok =
+      BinarySeek<DecodeKey>(seek_key, 0, num_restarts_, &index, comparator_);
 
   if (!ok) {
     return;
   }
-  SeekToRestartPoint(index);
 
   // Linear search (within restart block) for first key >= target
-  while (ParseNextDataKey<DecodeEntry>() && Compare(key_, seek_key) < 0) {
+  if (index == 0) {
+    SeekToRestartPoint(0);
+    if (ParseNextDataKey<DecodeEntry>() &&
+        global_seqno_ != kDisableGlobalSequenceNumber &&
+        Compare(key_, seek_key) < 0) {
+      // TODO(ajkr): Currently the `BinarySeek()` does not apply global seqnum
+      // to the data keys before comparing them to the seek key. If the seek key
+      // is for a snapshot seqnum then it's possible that the key at the
+      // `index`th restart is actually less than the seek key. So here we have
+      // to consider the case that the `index`th restart is smaller than seek
+      // key once global seqnum is applied. Since user keys in ingested files
+      // must be strictly ascending, we need to advance at most one position.
+      ParseNextDataKey<DecodeEntry>();
+    }
+  } else {
+    SeekToRestartPoint(index - 1);
```

Comment:
> This saves one key comparison in block seeks when the seek key does not
exactly equal the found key. This is always the case in data
blocks as we use a special value type for seek that is not present in
data keys, and almost always the case in index blocks.

Nevertheless, what would this branch do if restart point @ index is equal to target? Would this code scan through restart point @ index-1 just to figure out that restart point @ index is equal to the target, which we already figured out in BinarySeek but do not communicate to the caller? (Aside from seqno craziness...) Wouldn't it simplify calling code to have BinarySeek return whether 'index' is an exact or nearest-greater match?

Actually, wouldn't it be even simpler to add that exact vs. non-exact distinction to the old version of the code to optimize away the unnecessary comparison? (So there's no 0 vs. > 0 distinction)

Reply:
> Nevertheless, what would this branch do if restart point @ index is equal to target? Would this code scan through restart point @ index-1 just to figure out that restart point @ index is equal to the target

Yes, it sacrifices that case under the assumption it almost never happens.

> which we already figured out in BinarySeek but do not communicate to the caller? (Aside from seqno craziness...) Wouldn't it simplify calling code to have BinarySeek return whether 'index' is an exact or nearest-greater match?

Agreed. Then we don't have to rely on my unverified assumption about how often it happens.

> Actually, wouldn't it be even simpler to add that exact vs. non-exact distinction to the old version of the code to optimize away the unnecessary comparison? (So there's no 0 vs. > 0 distinction)

Will try it out now and see. It sounds right...

Reply:
> > Actually, wouldn't it be even simpler to add that exact vs. non-exact distinction to the old version of the code to optimize away the unnecessary comparison? (So there's no 0 vs. > 0 distinction)

> Will try it out now and see. It sounds right...

We still end up needing to expand the binary search range by one to distinguish the case where 0th restart key is less than, equal to, or greater than the seek key. Interestingly right now `BinarySeek()` never compares against the 0th restart key (which also explains why they didn't handle the edge case where an empty index block says it has one restart key but actually has zero). Anyways, I think the change ends up similar to this PR.

But still before proceeding with this PR I plan to (1) fix the seqno craziness so `BinarySeek()` actually lands in a consistent place regardless of global seqno/snapshot reads; and (2) change `BinarySeek()` to return whether it found an exact or non-exact match, and use that to skip comparing against restart keys.

Reply:
Done. I also redid the `BinarySeek()` to avoid changing the binary search. The number of key comparisons saved lessened as a result, but I am fine with leaving those for a future PR..